Initial implementation of numeric report charts.

martinpovolny opened the pull request:
Replacement for #3528 w/o use of "Consolidation"

Work towards numeric report charts: https://bugzilla.redhat.com/show_bug.cgi?id=1213294

martinpovolny commented:
![chart-cpus-1](https://cloud.githubusercontent.com/assets/51095/9306907/895136f4-44fb-11e5-80d8-8309599eeae5.png)
![chart-cpus-2](https://cloud.githubusercontent.com/assets/51095/9306908/896961de-44fb-11e5-87d3-96a72151232a.png)
![chart-cpus-3](https://cloud.githubusercontent.com/assets/51095/9306910/897102c2-44fb-11e5-95a3-ca94f24b8015.png)
![chart-cpus-4](https://cloud.githubusercontent.com/assets/51095/9306909/896d5140-44fb-11e5-88a7-9980b1a50d3b.png)

martinpovolny commented:
## Number of VMs per host

![chart-vms-1](https://cloud.githubusercontent.com/assets/51095/9307011/07af452c-44fc-11e5-8772-a0cfab0bd25a.png)
![chart-vms-2](https://cloud.githubusercontent.com/assets/51095/9307013/07b2f834-44fc-11e5-90d8-061542fc0c8b.png)
![chart-vms-3](https://cloud.githubusercontent.com/assets/51095/9307012/07b29d80-44fc-11e5-93df-5ad6f961600c.png)
![chart-vms-4](https://cloud.githubusercontent.com/assets/51095/9307014/07b6db98-44fc-11e5-87e4-fd178dad25f2.png)

dclarizio commented:
@martinpovolny This is looking pretty good . . . I did a report on disk space and noticed that maybe we could use the same formatting for the legend markers as we have in the report, as seen below:

**Report sort criteria**
![report sort](https://cloud.githubusercontent.com/assets/3433754/9318066/e7e6d4e6-44f4-11e5-8863-e32254d5ea2d.png)

**Report chart options**
![report chart](https://cloud.githubusercontent.com/assets/3433754/9318072/f04bd0d2-44f4-11e5-9de7-326951684b09.png)

**Note the legends with all the zeroes and no commas**
![bar chart](https://cloud.githubusercontent.com/assets/3433754/9317980/29e59450-44f4-11e5-888f-39f0e3687635.png)

**Report shows them as humanized, to MB, GB, TB, etc**
![bar chart 2](https://cloud.githubusercontent.com/assets/3433754/9317981/2c513e2e-44f4-11e5-81cf-1e22d0004c5b.png)

**Same in column view with vertical axis legends**
![col chart](https://cloud.githubusercontent.com/assets/3433754/9317982/2e715d92-44f4-11e5-9704-5513a0765e95.png)

This may be a jqPlots issue, but something to keep in mind for the future C3 charting is that with the flash charts, we could hover over a bar or chart item or node on a line chart and it would show a popup with the actual value.  Would be a nice feature to keep.

**Pie charts for values look good and they have hover text over the pie slices, but same note about formatting of the values shown**
![piechart](https://cloud.githubusercontent.com/assets/3433754/9318483/c45f7042-44f8-11e5-874c-130baefa3b30.png)

**Pie chart using counts didn't fare as well, chart is blank.**
![piecounts](https://cloud.githubusercontent.com/assets/3433754/9318520/0330562e-44f9-11e5-9776-4902a387b0bb.png)

**Column chart based on counts results in a line chart (strange)**
![chartcolcounts](https://cloud.githubusercontent.com/assets/3433754/9318587/c5109d80-44f9-11e5-9459-b011de295cb1.png)

![colcountchart](https://cloud.githubusercontent.com/assets/3433754/9318607/f052eb4c-44f9-11e5-97ec-461ffae38982.png)

dclarizio commented:
@martinpovolny Lots of feedback above, let's discuss where you're at tomorrow morning.  Thx, Dan

martinpovolny commented:
@dclarizio :
  1. fixed the missing chart and the line chart -- seems chart type information was missing in the json
  1. added a spec for 2 main cases of the count-based charts -- this shoul cover the case above
  1. fixed some problem in the report editor when switching chart types

dclarizio commented:
@martinpovolny more testing . . . 

**Summary tab set to counts with no aggregations set, Chart tab should not allow values.**  Currently lets you select values, but the data column has a null choice.  If no aggregations are being done, it should show Counts as a string with no choice and, of course, not show the column drop down.  Here are the screenshots:
![summary tab](https://cloud.githubusercontent.com/assets/3433754/9372002/6d1d846c-468e-11e5-9ebb-b7b3e87f3aca.png)
![charts tab](https://cloud.githubusercontent.com/assets/3433754/9372003/6f1df256-468e-11e5-890b-af06c6446614.png)

**Running the report with counts, not sure the output is correct.**  I can see some break lines in the report output and 1 of them lines up with the legend and has a value of 3 showing.  The others, I think the name is too long so I can't see the count in the legend???  The chart shows percentages (which is probably ok), but hovering the mouse over the pie slices doesn't tell me which provider it is or the count, there is no popup.  Basically, I can't tell if the chart matches the report.
![count chart report](https://cloud.githubusercontent.com/assets/3433754/9372094/973b83ce-468f-11e5-8372-fa2b51550151.png)

dclarizio commented:
@martinpovolny Simple test, ran the "Vendor and Guest OS" report with the stacked bar chart (so 3d charts by count) and the charts look identical, so it seems those are working as before.  Thought I would post some good news as well.  :)

**Before:**
![vendorandguest before](https://cloud.githubusercontent.com/assets/3433754/9372176/5d25a920-4690-11e5-8077-9be73608f6b7.png)

**After:**
![vendorandguest after](https://cloud.githubusercontent.com/assets/3433754/9372179/60399874-4690-11e5-974f-7a97252a2bcc.png)

dclarizio commented:
@martinpovolny More good news, testing of column charts with values.  Only issue here is from before, Y axis labels need formatting.  Would be nice to have hover over values, but we can wait for C3.

**Chart configuration:**
![colchartsummary](https://cloud.githubusercontent.com/assets/3433754/9372248/56981330-4691-11e5-992c-d2c2d88317ce.png)
![colchart charts](https://cloud.githubusercontent.com/assets/3433754/9372250/582894e0-4691-11e5-81d4-eaf3591e18ef.png)

**Chart/Report:**
![colchart report](https://cloud.githubusercontent.com/assets/3433754/9372273/9b8aabb0-4691-11e5-8747-ea60127e82f0.png)

dclarizio commented:
@martinpovolny Same issues as above for bar charts (X axis labels crazy) . . . they look pretty good tho:

![barchart](https://cloud.githubusercontent.com/assets/3433754/9372320/1849d8ba-4692-11e5-8e4b-a5e6ef86b7ba.png)

martinpovolny commented:
@dclarizio : Thinking about the use of `MiqReport::Formatting` for the formatting of the axis labels. The problem is that the ticks for numeric values are calculated by the javascript library. The same most likely holds for the flash charts. 

For one we don't know the values for the ticks when calculation the chart. Then depending on type of chart we might have a problem with associating labels of the axis with the values in the chart. In jqplot you can have a formatter class for the ticks or you can use a special axis renderer for a couple of typical axis such as Date and Cathegory.

Seems to me that the proper way would be providing formatters based on column types similarly to what we have in `db/fixtures/miq_report_formats.yml` -- bytes, mHz, percent etc. No idea how to do this for the flash charts. Also might be work investigating what we will do if we use the C3 charts instead and do in in a general way. Definitely even though the labels are pretty ugly (I can rotate them at least to fit better) I would say that we need a separate issue and some research for this.

@himdel : do you know that ways one can use in the C3 to format axis values?

himdel commented:
@martinpovolny sure, I think [this](http://c3js.org/reference.html#axis-x-tick-format) should be useable :)

martinpovolny commented:
@himdel : http://www.jqplot.com/docs/files/plugins/jqplot-canvasAxisTickRenderer-js.html#$.jqplot.CanvasAxisTickRenderer.formatter

so we need 
  1. some javascript code that would take understand how to format according to the `db/fixtures/miq_report_formats.yml` formats and 
  1. then we just pass in the format of the column that is being projected on the axis along with the other options as we generate the charts

@dclarizio : I'd probably create an issue for that

martinpovolny commented:
@himdel : do you know about some nice js library we could use? https://github.com/adamwdraper/Numeral-js maybe?

himdel commented:
Uh, looking at the formats, angular actually has most of these as filters but it'd be rather hacky to use here I guess... I've never heard of Numeral.js but looks like it could be useful for the numeric formats. We already have moment.js so we can use that for dates/times, and possibly something like sprintf-js to cover the rest?

I can look more into it tomorrow .. sounds like simply reimplementing most of miq_report/formatting.rb in JS, right?

martinpovolny commented:
@himdel : seems to me to be the best way to reimplement that in JS. In the reports we way for the user to specify his/her format so we might actually end really reimplementing it so that we have the same format specification on both sides.

martinpovolny commented:
@himdel : if you can look into it, I'd be glad. I can make the charts pass in the formating information for the right column, but If you could help with the javascript implementation of the formating function, it would be great help for me. Your experience with variety of javascript libraries would be usefull here ;-)

himdel commented:
@martinpovolny sure, no problem :) I'm assuming we want to keep the same function names mentioned in `:function` `:name`, I'll put it under .. say, `ManageIQ.charts.formatters` or something like that, ok?

himdel commented:
Support for strftime-like date format specifiers in `format_date` will depend on [this moment-strftime PR]( https://github.com/benjaminoakes/moment-strftime/pull/14 ). Is there any chance we may need other format specifiers not mentioned in the yaml?

dclarizio commented:
@martinpovolny Minor detail in pie popup, not sure why it shows 2 values with : separator.  Maybe legend has the value in the text, then it is adding the actual value again after the text in the popup.
![piepopup](https://cloud.githubusercontent.com/assets/3433754/9419784/a726e06e-4813-11e5-9518-495bb97eeca5.png)

martinpovolny commented:
## 2dim usecase
![report-chart-dim2-1](https://cloud.githubusercontent.com/assets/51095/9521288/4146a166-4cce-11e5-9686-0164db454e46.png)
![report-chart-dim2-2](https://cloud.githubusercontent.com/assets/51095/9521289/414d0538-4cce-11e5-8b50-24b8c854f059.png)
![report-chart-dim2-3](https://cloud.githubusercontent.com/assets/51095/9521291/41504554-4cce-11e5-9628-7f7030c6b965.png)
![report-chart-dim2-4](https://cloud.githubusercontent.com/assets/51095/9521290/414fb9d6-4cce-11e5-9a07-3ed49e36c6fa.png)
![report-chart-dim2-5](https://cloud.githubusercontent.com/assets/51095/9521292/4152a72c-4cce-11e5-805a-7b86fc282631.png)

martinpovolny commented:
The text shortening code is suboptimal at best. It simply cuts what's over certain limit `N` and adds "...".
We should probably change it to shortening what's longer than `N+3` and then cut at `N` and add the "...".

martinpovolny commented:
@dclarizio: fixed the "2 values with : separator..." in the last commit.

matthewd commented on lib/report_formatter/chart_common.rb:
```diff
@@ -2,7 +2,7 @@ module ReportFormatter
     def slice_legend(string, limit = LEGEND_LENGTH)
       string = string.to_s
-      string.length > limit ? string.slice(0, limit) + "..." : string
+      string = string.length > limit ? string.slice(0, limit) + "..." : string
```
http://api.rubyonrails.org/classes/String.html#method-i-truncate ?

martinpovolny replied:
Thx, fixed.

tenderlove commented on lib/report_formatter/chart_common.rb:
```diff
@@ -347,13 +346,166 @@ def build_reporting_chart_dim2
+
+      # For each value in sort2 column we create a series.
+      sort2_values = mri.table.data.each_with_object({}) { |row, h| h[row[sort2]] = true }
+      sort2_values.each_key do |val2|
```
Could this be:

```ruby
sort2_values = mri.table.data.map { |row| row[sort2] }.uniq
sort2_values.each do |val2|
```

martinpovolny replied:
Yes! Thx

dclarizio commented:
@martinpovolny Tried a 2 dim chart, got the following when viewing the report/chart:

I just noticed when taking the screen shots that on Summary tab I specified Counts, so on Charts tab it shows "Counts", but I have the Avg aggregation selected.  I'm thinking maybe that's an editor bug?  I will change it to use an aggregation on the Summary tab and the same on Charts tab to see if that works, but this still may be a bug in editor where it should have used Counts and not showed me the aggregation drop down.
```
[----] I, [2015-08-28T14:21:30.039953 #26232:e4d98c]  INFO -- : Started GET "/report/render_chart?height=250&rand=431984952&width=350" for 127.0.0.1 at 2015-08-28 14:21:30 -0700
[----] I, [2015-08-28T14:21:30.100082 #26232:e4d98c]  INFO -- : Processing by ReportController#render_chart as JSON
[----] I, [2015-08-28T14:21:30.100291 #26232:e4d98c]  INFO -- :   Parameters: {"height"=>"250", "rand"=>"431984952", "width"=>"350"}
[----] F, [2015-08-28T14:21:30.276852 #26232:e4d98c] FATAL -- : Error caught: [ArgumentError] comparison of String with nil failed
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:448:in `sort'
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:448:in `build_numeric_chart_grouped_2dim'
/home/dclarizio/dev/manageiq/lib/report_formatter/jqplot.rb:142:in `build_numeric_chart_grouped_2dim'
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:560:in `build_reporting_chart_numeric'
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:62:in `call'
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:62:in `build_document_body'
```
Report details:
![charts_summary](https://cloud.githubusercontent.com/assets/3433754/9557406/a258f450-4d90-11e5-99f8-ea7e05d3ae38.png)
![charts_charts](https://cloud.githubusercontent.com/assets/3433754/9557407/a4379ae2-4d90-11e5-82c0-54f404d698bc.png)

dclarizio commented:
@martinpovolny Same error as above with Summary/Charts configured correctly:
![summary1](https://cloud.githubusercontent.com/assets/3433754/9557557/e3489abe-4d91-11e5-85d5-464538e7d952.png)
![charts1](https://cloud.githubusercontent.com/assets/3433754/9557561/e4ef4034-4d91-11e5-93e1-ede7a4428350.png)

dclarizio commented:
@martinpovolny Changing Summary and Charts tabs to use Counts instead of values as above, charts are drawn successfully:
![summary2](https://cloud.githubusercontent.com/assets/3433754/9557757/5ee57984-4d93-11e5-93ed-e281f11a8dcf.png)
![charts2](https://cloud.githubusercontent.com/assets/3433754/9557759/6127d412-4d93-11e5-95ab-72dc48988bce.png)
![report2](https://cloud.githubusercontent.com/assets/3433754/9557753/5ac8bdf2-4d93-11e5-8fee-32aa6f0fe13e.png)

dclarizio commented:
@martinpovolny I just noticed something in my prior comment (and screen shots).  I selected the Bars (2D) chart, but it is creating Columns Stacked (2D) charts.  Changing to Columns (2D) chart also shows the same thing, Columns Stacked (2D).  Changing to Bars Stacked (2D), same thing . . . almost like it's hard coded to do Columns Stacked.

Donuts and Pies seem to work fine.

martinpovolny commented:
1. fixed report editor (disallow value charts when summary changed from "Yes")
  1. fixed field parsing in charts to fix above reported `lib/report_formatter/chart_common.rb:448:in `sort'` happening for fields with underscore

dclarizio commented:
@martinpovolny So, the Bars (2D) and Columns (2D) charts CAN support multiple series, as shown in the sample charts in the editor.  It simply places the series values for each axis point next to one another:
![bars2d](https://cloud.githubusercontent.com/assets/3433754/9636916/d4aca59e-5152-11e5-8a90-839df46007af.png)
![charts2d](https://cloud.githubusercontent.com/assets/3433754/9636919/d76e307c-5152-11e5-858f-261af01cc7bc.png)

Running these reports incorrectly results in a change to the stacked charts:
![bars2dreport](https://cloud.githubusercontent.com/assets/3433754/9636976/1ee7b388-5153-11e5-9bb0-cf134eb24005.png)
![cols2dreport](https://cloud.githubusercontent.com/assets/3433754/9636979/21d4ae2a-5153-11e5-81a0-475846cebbcd.png)

Along these lines, I think we can also remove `(2D)` from all of the chart type names as none of them support 3D as DHTMLX used to.

dclarizio commented:
@martinpovolny Noticed the values chooser is using the internal field name, should show the same as in summary tab.
![summarycalcs](https://cloud.githubusercontent.com/assets/3433754/9638162/d8e64e38-5159-11e5-8683-6432a7c71d2c.png)
![chartsvalues](https://cloud.githubusercontent.com/assets/3433754/9638163/dc2a9b26-5159-11e5-8648-ee286d0d30d6.png)

dclarizio commented:
@martinpovolny When linked to productization, run this report (screenshot) and when rendering the chart, I get this exception (log).  Was trying to see if the dhtmlx charts did the non-stacked bar/col charts properly still.  Just noticed this is the same error as above, so perhaps this code is also in productization?
![linkedchart](https://cloud.githubusercontent.com/assets/3433754/9638405/6fdeaee2-515b-11e5-9e55-2663c9065e54.png)
```
[----] I, [2015-09-02T10:11:39.560391 #19620:5d7988]  INFO -- : Started GET "/report/render_chart?height=250&rand=513429900&width=350" for 127.0.0.1 at 2015-09-02 10:11:39 -0700
[----] I, [2015-09-02T10:11:39.617529 #19620:5d7988]  INFO -- : Processing by ReportController#render_chart as HTML
[----] I, [2015-09-02T10:11:39.617761 #19620:5d7988]  INFO -- :   Parameters: {"height"=>"250", "rand"=>"513429900", "width"=>"350"}
[----] F, [2015-09-02T10:11:39.786693 #19620:5d7988] FATAL -- : Error caught: [ArgumentError] comparison of String with nil failed
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:448:in `sort'
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:448:in `build_numeric_chart_grouped_2dim'
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:560:in `build_reporting_chart_numeric'
```

martinpovolny commented:
@dclarizio : thanks for detailed testing!

Now I:
  1. fixed the field labels in editor
  1. unified the formating of counts and value based 2dim charts

dclarizio commented:
@martinpovolny looks like a relevant spec test failure on vmdb.

dclarizio commented:
@martinpovolny Reran my saved chart report (with 2 series and non-stacked chart selected) with the new code, got this error again when chart is rendered for display (same as above with productization):
![chartcols](https://cloud.githubusercontent.com/assets/3433754/9672207/517b5854-524d-11e5-896e-8828923c4d15.png)
```
[----] I, [2015-09-03T15:03:15.888138 #5574:ad9990]  INFO -- : Completed 200 OK in 500ms (Views: 0.3ms | ActiveRecord: 16.4ms)
[----] I, [2015-09-03T15:03:16.048822 #5574:ad9990]  INFO -- : Started GET "/report/render_chart?height=250&rand=558489395&width=350" for 127.0.0.1 at 2015-09-03 15:03:16 -0700
[----] I, [2015-09-03T15:03:16.109741 #5574:ad9990]  INFO -- : Processing by 
ReportController#render_chart as JSON
[----] I, [2015-09-03T15:03:16.109938 #5574:ad9990]  INFO -- :   Parameters: {"height"=>"250", "rand"=>"558489395", "width"=>"350"}
[----] F, [2015-09-03T15:03:16.274265 #5574:ad9990] FATAL -- : Error caught: [ArgumentError] comparison of String with nil failed
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:448:in `sort'
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:448:in `build_numeric_chart_grouped_2dim'
/home/dclarizio/dev/manageiq/lib/report_formatter/jqplot.rb:127:in `build_numeric_chart_grouped_2dim'
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:560:in `build_reporting_chart_numeric'
```

dclarizio commented:
@martinpovolny Easy fix, remove (2D) from chart type names since there are no 3D charts:
![charttype](https://cloud.githubusercontent.com/assets/3433754/9672307/5f6d2324-524e-11e5-8f5a-db4d81cc8585.png)

martinpovolny commented:
@dclarizio : on "Reran my saved chart report (with 2 series and non-stacked chart selected) with the new code, got this error again when chart is rendered for display (same as above with productization):"

Can  you, please, either create the report again from scratch or send me also a screenshot of the "Summary page". I cannot reproduce the error. 

Seems to me you must have only one sort criteria in the Summary (which the editor does not allow any more) -- so you have an invalid combination or I don't know. Just cannot create the situation on my box.

dclarizio commented:
@martinpovolny Still seeing the exception when rendering the values chart, screenshots below.  Also IRC'd you a link to my DB.
![errsummary](https://cloud.githubusercontent.com/assets/3433754/9690849/b9cad762-52f3-11e5-95b4-d101eef97e86.png)
![errcharts](https://cloud.githubusercontent.com/assets/3433754/9690850/bb955e28-52f3-11e5-93de-703e7e614db6.png)

martinpovolny commented:
@dclarizio : please, retest with the last patch. I have also modified the spec to fail w/o the change.

dclarizio commented:
@martinpovolny Exception is gone upstream.  When productized, I get the following when rendering the chart with values (counts renders fine).  Should be able to reproduce this with the DB I sent you.
```
[----] I, [2015-09-04T13:21:45.238407 #7786:64f988]  INFO -- : Started GET "/report/render_chart?height=250&rand=144678747&width=350" for 127.0.0.1 at 2015-09-04 13:21:45 -0700
[----] I, [2015-09-04T13:21:45.294838 #7786:64f988]  INFO -- : Processing by ReportController#render_chart as HTML
[----] I, [2015-09-04T13:21:45.295038 #7786:64f988]  INFO -- :   Parameters: {"height"=>"250", "rand"=>"144678747", "width"=>"350"}
[----] F, [2015-09-04T13:21:45.453925 #7786:64f988] FATAL -- : Error caught: [NoMethodError] undefined method `match' for nil:NilClass
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:437:in `build_numeric_chart_grouped_2dim'
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:560:in `build_reporting_chart_numeric'
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:62:in `call'
/home/dclarizio/dev/manageiq/lib/report_formatter/chart_common.rb:62:in `build_document_body'
```

martinpovolny commented:
@dclarizio : the above error seems as if you where loading a broken report from DB. That's a failure in parsing the saved chart column from the report definition.

Also I was not able to reproduce this even when I tried your database, opened "Dan's Chart Tests" and then tried to edit your report (changing Value based to Counts based and more).

dclarizio commented:
@martinpovolny OK, so I created a chart report from scratch when productized and it worked just fine.  I also copied the suspect chart and when I went to the summary screen, non of the aggregations were defined.  Defined those and the charts worked there as well.  Merging this now.

chessbyte commented on app/assets/javascripts/miq_jqplot.js:
```diff
@@ -68,6 +71,27 @@ function jqplot_redraw_charts() {
     }
 }
 
+function jqplot_pie_highligh_values(str, seriesIndex, pointIndex, plot) {
```
Should highligh be spelled highlight?

chessbyte commented on app/assets/javascripts/miq_jqplot.js:
```diff
@@ -68,6 +71,27 @@ function jqplot_redraw_charts() {
+           plot.series[seriesIndex].data[pointIndex][1];
+}
+
+function jqplot_pie_highligh(str, seriesIndex, pointIndex, plot) {
```
Should highligh be spelled highlight?

martinpovolny replied:
Yes. I'll look into it.

Also In master we are scoping all the javascript functions under `ManageIQ.whatever` context. So I will have to touch this area soon. But we do not have the scoping in production so I have to backport this first.

Also I still have to use @himdel #4003 and axis ticks formating.
